### ISSUE TITLE
Node controller kills pods on a recently recovered node

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#13258]
This test is failing pretty frequently at head. It fails with the following message:

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:567
Expected error:
    <*errors.errorString | 0xc2088f40b0>: {
        s: "failed to wait for pods responding: pod with UID 3a70b8fa-4ca9-11e5-8a57-42010af00443 is no longer a member of the replica set.  Must have been restarted for some reason.  Current replica set: &{{ } {/api/v1/namespaces/e2e-tests-resize-nodes-ag64m/pods 16793} []}",
    }
    failed to wait for pods responding: pod with UID 3a70b8fa-4ca9-11e5-8a57-42010af00443 is no longer a member of the replica set.  Must have been restarted for some reason.  Current replica set: &{{ } {/api/v1/namespaces/e2e-tests-resize-nodes-ag64m/pods 16793} []}
not to have occurred

@kubernetes/goog-testing @quinton-hoole @piosz @jszczepkowski

**Comment by jszczepkowski** [comment]
The is a problem in the following scenario:
1. A node is removed due to a network partition.
2. The network partition is over, the node becomes ready again.
3. (immediately after 3) A pod is created on the node.
4. Node controllers kills the pod with "reason": "NodeControllerEviction".
The eviction in point 4. is incorrect. There seems to be a bug in node controller.

The whole event of the eviction:

{ "metadata": { "name": "additionalpod.13ffd324adb79a6b", "namespace": "e2e-tests-resize-nodes-2dd5q", "selfLink": "/api/v1/namespaces/e2e-tests-resize-nodes-2dd5q/events/additionalpod.13ffd324adb79a6b", "uid": "716815f5-5092-11e5-abd1-42010af07472", "resourceVersion": "3435", "creationTimestamp": "2015-09-01T10:16:00Z", "deletionTimestamp": "2015-09-01T11:16:00Z" }, "involvedObject": { "kind": "Pod", "namespace": "e2e-tests-resize-nodes-2dd5q", "name": "additionalpod", "uid": "6d8785e4-5092-11e5-abd1-42010af07472", "apiVersion": "v1", "resourceVersion": "3431" }, "reason": "NodeControllerEviction", "message": "Deleting Pod additionalpod from Node e2e-test-jsz-minion-c909", "source": { "component": "controllermanager" }, "firstTimestamp": "2015-09-01T10:16:00Z", "lastTimestamp": "2015-09-01T10:16:00Z", "count": 1 }


The test is flaky in 50% of cases, so, it can be easily reproduced on a private e2e cluster.

**Comment by jszczepkowski** [comment]
@smarterclayton

**Comment by wojtek-t** [comment]
Nice debugging @jszczepkowski 
@gmarek - it would be great if you could fix that soon

**Comment by gmarek** [comment]
I looked at the code, and I was able to find one problem: because `Accept` in https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/rate_limited_queue.go#L172 is blocking, we might miss the value removal, which would prevent eviction/termination. Because of that if we re-add Node when already waiting on 'Accept' we will remove everything from it (including new stuff). 

I'm not really sure why we want to NodeController queues to be 'non-leaky' (i.e. block on Accept) instead of doing non-blocking check, and if failed try again later. @smarterclayton 

In addition, while talking with @davidopp this morning, I couldn't recall why exactly we want to rate limit Pod removal and not only guard against evicting too many Nodes in the same time.

**Comment by gmarek** [comment]
Ping @smarterclayton - can you give any input on this? I'm going to debug this in free cycles (which may appear in not-too-distant future or not), so any help with this will be appreciated.

**Comment by smarterclayton** [comment]
It's entirely possible it's not getting flushed from the cleanup queue due
to a bug.  I will look when I have time tonight.

On Sep 3, 2015, at 6:52 AM, Marek Grabowski <notifications@github.com>
wrote:

Ping @smarterclayton <https://github.com/smarterclayton> - can you give any
input on this? I'm going to debug this in free cycles (which may appear in
not-too-distant future or not), so any help with this will be appreciated.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13258#issuecomment-137406920>
.

**Comment by gmarek** [comment]
I found following races:
- waiting in callback in terminationEvictor until RateLimitter returns true, we loose 'cancelPodEviction' call if happens then
- cancelPodEviction is not atomic and is not synchronized with callbacks, e.g. we may get 'cancelPodEviction' call when podEvictor's callback is running, 'cancel...' will finish, and at the end of the callback we'll add a value to the terminationEvictor

For the first one I just removed 'leaky' from the queue (I'm not sure what was the reasoning behind this), as for the second one I put all calls in a single mutex.

Both of them are unlikely reason behind the test failures and I suspect that the main culprit is the Watch latency: we observe Node readiness in the test, before it's delivered to the NodeController, and we create a Pod assigned to the Node straight away, which in turn gets evicted before Node becomes ready. I added additional wait in the test, which should help with this issue there, though this kind of race can happen in production (though is very unlikely). 

Hopefully #13642 will fix this

@smarterclayton @davidopp

**Comment by smarterclayton** [comment]
Those are definitely races - I will try to look tomorrow when I'm back in
the office but what you describe as fixes seems correct to me.

On Sep 3, 2015, at 6:52 AM, Marek Grabowski <notifications@github.com>
wrote:

Ping @smarterclayton <https://github.com/smarterclayton> - can you give any
input on this? I'm going to debug this in free cycles (which may appear in
not-too-distant future or not), so any help with this will be appreciated.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/13258#issuecomment-137406920>
.

**Comment by quinton-hoole** [comment]
The test originally reported in this issue is still flaky.  Re-opening.

Failed 3 times in the last 30 runs. Stability: 90 %

** Kubernetes e2e suite.Nodes Network when a minion node becomes unreachable [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster ** 

```
Stacktrace

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/resize_nodes.go:549
Expected error:
    <*errors.errorString | 0xc20802bf70>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
not to have occurred
```

The essence of the problem seems to be here:

```
Sep 10 12:31:05.062: INFO: [jenkins@130.211.166.128:22] Running    `sudo iptables --insert OUTPUT --destination e2e-flaky-master --jump DROP`
...
Sep 10 12:31:05.062: INFO: Waiting 2m0s for node e2e-flaky-minion-50ti to be not ready after simulated network failure
...
Sep 10 12:31:45.097: INFO: Node e2e-flaky-minion-50ti condition 1/1: type: Ready, status: Unknown, reason: "Kubelet stopped posting node status.", message: "", last transition time: 2015-09-10 12:31:41 -0700 PDT
Sep 10 12:31:45.097: INFO: Successfully found node e2e-flaky-minion-50ti readiness to be false
Sep 10 12:31:45.097: INFO: Waiting for pod my-hostname-net-a0j4y to be removed
Sep 10 12:32:05.098: INFO: Waiting for pod my-hostname-net-a0j4y to disappear
Sep 10 12:32:05.132: INFO: Pod my-hostname-net-a0j4y still exists
...
Sep 10 12:36:45.109: INFO: Pod my-hostname-net-a0j4y still exists

quinton-hoole: [implicit] Pod my-hostname-net-a0j4y did not disappear within the 5 minute timeout.

quinton-hoole: [implicit] what follows is just cleanup.  The above is the actual failure.

STEP: Unblock network traffic from node e2e-flaky-minion-50ti to the master
Sep 10 12:36:45.394: INFO: [jenkins@130.211.166.128:22] Running    `sudo iptables --delete OUTPUT --destination e2e-flaky-master --jump DROP`
Sep 10 12:36:45.394: INFO: [jenkins@130.211.166.128:22] stdout:    ""
Sep 10 12:36:45.394: INFO: [jenkins@130.211.166.128:22] stderr:    ""
Sep 10 12:36:45.394: INFO: [jenkins@130.211.166.128:22] exit code: 0
Sep 10 12:36:45.394: INFO: [jenkins@130.211.166.128:22] error:     <nil>
```

The relevant code is here:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/resize_nodes.go#L375

The implicit 5 minute timeout is here:

https://github.com/kubernetes/kubernetes/blob/a48f218264f62267ea743326d6600ad98951c6cf/test/e2e/util.go#L646

**Comment by quinton-hoole** [comment]
It's worth noting that the failure mode here is that pods on failed nodes are failing to be removed by the system, as opposed to new pods being erroneously deleted, which has been the focus of this debugging effort thus far.  Perhaps we should spawn a separate issue to debug this second failure mode?  @gmarek @jszczepkowski  ?

**Comment by quinton-hoole** [comment]
I can trace the problem back to at least Sep 3, 2015 3:41:03 PM. That's as far back as our history on our soak test cluster extends.  The problem probably predates that substantially.

**Comment by quinton-hoole** [comment]
I would imagine that this is pretty simple to reproduce on a local test cluster.  Just build one, and run the e2e suite against it a few times.  What we're seeing is typically two successful e2e runs, followed by consistent failures of the above tests for the reasons stated.

**Comment by gmarek** [comment]
Thanks for bringing this up, I actually want to discuss it with you. To my understanding it's a completely different problem, namely we do have 5 minute grace period for Pods to get evicted. I was considering lowering this period for our test clusters, as I don't think it's worth to be so conservative. WDYT about setting the pod-eviction-period to, say 2 minutes. This should both unflake and speed up this test, and I don't think it'll cause any problems.

**Comment by quinton-hoole** [comment]
@gmarek reducing the pod eviction timeout for the tests seems perfectly reasonable.  Perhaps even for production?  Any idea where the current 5 minute pod eviction came from?  Science or magic? :-)

Oh, and by the way, you can ignore my last comment (which I've just deleted "pretty simple to reproduce)"). It was intended for a different issue, but my cut 'n paste finger failed ;-)

**Comment by gmarek** [comment]
It's a science-ish. We don't want a isolated Kubelets crash to cause Pods eviction. The stuff that happens around that were: marking Node as unhealthy takes ~40s, `monit` check period was 2m (I believe we moved to systemd, but I wasn't paying attention to this effort). Kubelet start time was up to ~40s last time I checked (it was a month ago or sth).  Altogether it takes ~3m20s, so it should be safe to set lower the time to 4m, but no one thought it's worth the effort:)

For testing I don't think we care as much - if Kubelet fails unexpectedly we're likely to fail test either way.

**Comment by quinton-hoole** [comment]
See also my recent comments in #13430 ("Re-evaluate rate limiting parameters").

@gmarek Thanks for your fantastically lucid explanation.  Makes sense.  Lets drop the pod eviction timeout to double the node marking interval (i.e. 2 x 40s ?) for e2e tests, on the basis that we don't need to accommodate crashing kubelets in the test.

As an aside, and as a separate task, I think that we should explicitly express the pod eviction timeout in production as a function of those input variables that you mention, and update it when those variables change.

**Comment by gmarek** [comment]
There's a slight problem of setting a sane value for Kubelet start time, as it has huge variance:/ I'll send a PR for changing the grace period in e2e clusters.

**Comment by gmarek** [comment]
cc @dchen1107 to fill in the detail of current state of Kubelet restart/startup.

**Comment by gmarek** [comment]
Lowering the priority, as this was mostly fixed. 

Ping @dchen1107 on Kubelet startup time and restart period.

**Comment by davidopp** [comment]
I'm going to move this out of 1.1.

@dchen1107 would be good to know current Kubelet startup time and monitor/systemd/whatever liveness check interval for kubelet.

**Comment by gmarek** [comment]
I haven't seen this kind of failure for quite some time. Reopen if it'll reappear.
